x/poe/keeper/keeper.go

```diff
@@ -40,29 +42,43 @@ func NewKeeper(
 	}
 
 	return Keeper{
-		codec:       marshaler,
-		storeKey:    key,
-		paramStore:  paramSpace,
-		twasmKeeper: twasmK,
+		codec:             marshaler,
+		storeKey:          key,
+		paramStore:        paramSpace,
+		twasmKeeper:       twasmK,
+		contractAddrCache: make(map[types.PoEContractType]sdk.AccAddress),
 	}
 }
 
 // SetPoEContractAddress stores the contract address for the given type. If one exists already then it is overwritten.
 func (k Keeper) SetPoEContractAddress(ctx sdk.Context, ctype types.PoEContractType, contractAddr sdk.AccAddress) {
 	store := ctx.KVStore(k.storeKey)
 	store.Set(poeContractAddressKey(ctype), contractAddr.Bytes())
+	k.contractAddrCache[ctype] = contractAddr
 }
 
 // GetPoEContractAddress get the stored contract address for the given type
 func (k Keeper) GetPoEContractAddress(ctx sdk.Context, ctype types.PoEContractType) (sdk.AccAddress, error) {
 	if err := ctype.ValidateBasic(); err != nil {
 		return nil, sdkerrors.Wrap(err, "contract type")
 	}
+
+	// try to get addr from cache
+	addr := k.contractAddrCache[ctype]
+	if addr != nil || len(addr) != 0 {
+		return addr, nil
+	}
+
+	// if not in cache, try to get addr from store
```

Comment:
This path should never happen. We can either fail or return from the storage.
No need to put the address into the cache. There may be a race condition otherwise